### ISSUE TITLE
[RFC]: Sparse-attention-based self-speculative decoding (StreamingLLM, then Vegas)

### ISSUE THREAD

**Issue by npz7yyk** [vllm-project/vllm#47351]
### Motivation.

Long-context decoding is typically bottlenecked by the KV cache access. Self-speculative decoding with sparse attention [1] drafts tokens using the target model over a subset of the KV cache, then verifies them in parallel against the full cache. This delivers a lossless speedup with no separate draft model, no extra weights, and no training.

vLLM's spec-decode framework currently covers ngram, Eagle, Medusa, and MTP. These proposers either require a draft model or rely on n-gram matching. None supports a same-model, sparse-KV draft. This RFC adds one.

I am a co-author of Vegas [2]. While developing Vegas, I implemented self-speculative decoding with sparse attention in vLLM, with two sparse attention algorithms for drafting: StreamingLLM (sliding window with attention sink) [3] and Vegas. Measured as accumulated decoding throughput on one H100 with Qwen3-8B, replaying AIME 2025 [4] prompts 32 times: vanilla vLLM reaches ~1,000 tokens/s; the StreamingLLM drafter (`num_speculative_tokens=4`) reaches ~1,600 tokens/s; Vegas (`num_speculative_tokens=6`) reaches ~1,850 tokens/s. More evaluation results are available in the Vegas [paper](https://arxiv.org/abs/2602.07223).

A runnable implementation built on vLLM v0.16.0 with benchmark is public at https://github.com/platformxlab/vegas ([feature commit](https://github.com/platformxlab/vegas/commit/f49fd737a22c2e36a461600e85cc24f6177c27af)). I'd like to upstream it in two PRs.

[1] MagicDec: Breaking the Latency-Throughput Tradeoff for Long Context Generation with Speculative Decoding [arxiv:2408.11049](https://arxiv.org/abs/2408.11049)  
[2] Vegas: Self-Speculative Decoding with Verification-Guided Sparse Attention [arXiv:2602.07223](https://arxiv.org/abs/2602.07223)  
[3] Efficient Long-Context Generation with Sliding Window and Attention Sink [arXiv:2309.17453](https://arxiv.org/abs/2309.17453)  
[4] American Invitational Mathematics Examination (AIME) 2025 [`math-ai/aime25`](https://huggingface.co/datasets/math-ai/aime25)

### Proposed Change.

Add a new speculative method `method="sparse_attn"` with a `SparseAttnProposer` that reuses the target model to draft under a sparse KV pattern, selected by a `sparse_attn_algorithm` knob.

- **PR 1:** the proposer framework + **StreamingLLM** as the default sparse pattern. A runnable branch is available at [`npz7yyk/vegas@vllm-pr1`](https://github.com/npz7yyk/vegas/tree/vllm-pr1), based on vLLM [`releases/v0.24.0`](https://github.com/vllm-project/vllm/tree/releases/v0.24.0) (commit [`ee0da84ab`](https://github.com/vllm-project/vllm/commit/ee0da84ab)), and it forward-ports cleanly to `main`; a PR is thus ready to be submitted once the approach is confirmed here.
- **PR 2:** **Vegas** algorithm (verification-guided KV selection). This needs a small flash-attention change to expose per-token attention logits from the verify pass as a byproduct; I'll propose that separately against the FA repo.

New config fields on `SpeculativeConfig`:

| Field | Meaning | Default |
| --- | --- | --- |
| `sparse_attn_algorithm` | sparse KV pattern (`"streamingllm"`) | `"streamingllm"` |
| `sparse_attn_ratio` | fraction of KV kept for drafting | `0.05` |
| `sparse_attn_min_tokens` | floor on per-request KV budget | `256` |

### PR 1: framework + StreamingLLM

New self-contained module:

```text
vllm/v1/spec_decode/sparse_attn/
├── __init__.py                     # exports SparseAttnProposer
├── proposer.py                     # SparseAttnProposer: drives the self-speculative draft loop
└── attn_overrider/
    ├── __init__.py                 # BaseAttnOverrider + build_attention_overrider() dispatch
    └── streamingllm.py             # StreamingLLMAttnOverrider: sink + sliding-window pattern
```

Wiring into existing vLLM files:

| File | What it does |
| --- | --- |
| `vllm/config/speculative.py` | Adds `method="sparse_attn"` + the `sparse_attn_*` fields |
| `vllm/v1/worker/gpu_model_runner.py` | Constructs and drives `SparseAttnProposer` |
| `vllm/v1/core/sched/scheduler.py` | Reserves lookahead slots so KV pages are allocated for draft tokens |

The change is a self-contained new module plus small, localized wiring edits (config + scheduler lookahead + model-runner dispatch); it does not touch the existing spec-decode proposers.

### PR 2: Vegas

Adds one algorithm slot + its kernels, and swaps in the flash-attention build that exposes per-token attention scores:

```text
vllm/v1/spec_decode/sparse_attn/attn_overrider/
├── vegas.py                        # VegasAttnOverrider: verification-guided KV selection
└── utils/
    ├── __init__.py
    ├── varlen_reduce.py            # CUDA kernel: reduce per-query scores -> per-token metric
    └── varlen_topk.py              # CUDA kernel: variable-length top-k KV selection
```

Plus small edits:

| File | Change |
| --- | --- |
| `attn_overrider/__init__.py` | Register the `"vegas"` branch in the dispatch |
| `vllm/config/speculative.py` | Add `"vegas"` to the `sparse_attn_algorithm` choices |
| companion flash-attention PR | Expose per-token logits via a `scores` output on `flash_attn_varlen_func` (separate PR against the FA repo) |

### Potential Concern

To keep `SparseAttnProposer` self-contained, the sparse KV pattern is currently applied by wrapping the attention backend at runtime during the draft pass (FlashAttention for now). It works, but a runtime wrap of core attention is a design choice I'd like input on: is this acceptable, or would maintainers prefer a more formal override mechanism in the attention layer? Happy to align with whatever fits vLLM's attention abstraction best.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by HaizhouPeng** [comment]
+++++
